### PR TITLE
#5301 sensitive fields acl

### PR DESCRIPTION
**Pull request by awgeorge**

### CODE REVIEW

**Comment by awgeorge** [comment]
Closes: #5301

**Comment by flovilmart** [comment]
Nice one! I’ll review shortly! 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

File: src/Controllers/DatabaseController.js [hunk 1]
```diff
@@ -1390,6 +1410,42 @@ class DatabaseController {
...
+      Object.keys(query).length === 0 &&
+      auth &&
+      auth.user &&
+      aclGroup.indexOf(auth.user.id) > -1
+    )
+      return null;
+
+    let protectedKeys;
+    [...(auth.userRoles || []), '*'].forEach(role => {
+      // If you are in multiple groups assign the role with the least protectedKeys.
```
Comment (milesrichardson) [inline src/Controllers/DatabaseController.js]: Good call on the warning. Length comparison seems dangerous. Idea: start with the base assumption that all protectedKeys are shielded, then iterate through each role, and each item in protectedKeys. For any fields that are in `protectedKeys`, but not `protectedFields[role]`, remove the key from the return value of `protectedKeys`

Comment (awgeorge) [reply src/Controllers/DatabaseController.js]: Yeah - Much better idea. Thanks! Implemented in next commit. Hope your project works out.

**Comment by flovilmart** [comment]
@acinader @dplewis can you give a review to this one? I believe the community is aching for those features.

**Comment by acinader** [comment]
yes, I will.

File: src/Controllers/DatabaseController.js [hunk 2]
```diff
@@ -162,7 +162,15 @@ const validateQuery = (query: any): void => {
...
 // Filters out any data that shouldn't be on this REST-formatted object.
-const filterSensitiveData = (isMaster, aclGroup, className, object) => {
+const filterSensitiveData = (
+  isMaster,
+  aclGroup,
+  className,
+  protectedFields,
+  object
+) => {
+  protectedFields && protectedFields.forEach(k => delete object[k]);
```
Comment (acinader) [inline src/Controllers/DatabaseController.js]: nice.  This function is now changed to not operate on user alone.  it'll strip sensitive fields for all object.  Got it.  nice!

File: src/Controllers/DatabaseController.js [hunk 3]
```diff
@@ -1390,6 +1410,42 @@ class DatabaseController {
...
+      return null;
+
+    let protectedKeys = Object.values(protectedFields).reduce(
+      (acc, val) => acc.concat(val),
+      []
+    ); //.flat();
+    [...(auth.userRoles || [])].forEach(role => {
+      const fields = protectedFields[role];
+      if (fields) {
+        protectedKeys = protectedKeys.filter(v => fields.includes(v));
```
Comment (acinader) [inline src/Controllers/DatabaseController.js]: `protectedKeys.filter(fields.includes)`, right?

File: src/Controllers/DatabaseController.js [hunk 4]
```diff
@@ -1390,6 +1410,42 @@ class DatabaseController {
     }
   }
 
+  addProtectedFields(
```
Comment (acinader) [inline src/Controllers/DatabaseController.js]: Confusing name to me.  This is returning an array of protected fields for  `auth` of this request..., right?

File: src/ParseServer.js [hunk 5]
```diff
@@ -343,14 +345,40 @@ function injectDefaults(options: ParseServerOptions) {
...
+      );
+
+    const userSensitiveFields = Array.from(
+      new Set([
+        ...(defaults.userSensitiveFields || []),
+        ...(options.userSensitiveFields || []),
+      ])
+    );
+
+    /* eslint-enable no-console */
```
Comment (acinader) [inline src/ParseServer.js]: move up to 355

File: spec/UserPII.spec.js [hunk 6]
```diff
@@ -521,5 +521,699 @@ describe('Personally Identifiable Information', () => {
...
+    });
+  });
+
+  describe('with configured sensitive fields via CLP', () => {
+    beforeEach(done => {
+      reconfigureServer({
+        protectedFields: {
+          _User: { '*': ['ssn', 'zip'], 'role:Administrator': [] },
+        },
+      }).then(() => done());
```
Comment (acinader) [inline spec/UserPII.spec.js]: `.then(done)`

File: spec/UserPII.spec.js [hunk 7]
```diff
@@ -521,5 +521,699 @@ describe('Personally Identifiable Information', () => {
...
+      const userObj = new (Parse.Object.extend(Parse.User))();
+      userObj.id = user.id;
+      userObj.fetch().then(
+        fetchedUser => {
+          expect(fetchedUser.get('email')).toBe(EMAIL);
+          expect(fetchedUser.get('zip')).toBe(ZIP);
+          expect(fetchedUser.get('ssn')).toBe(SSN);
+          done();
+        },
+        e => done.fail(e)
```
Comment (acinader) [inline spec/UserPII.spec.js]: `, done.fail`

File: spec/Schema.spec.js [hunk 8]
```diff
@@ -653,6 +656,68 @@ describe('SchemaController', () => {
       });
   });
 
+  it('refuses to add CLP with incorrect find', done => {
```
[the remark this answers is not in the document]
Comment (awgeorge) [reply spec/Schema.spec.js]: Yeah - no idea what I was up to here. It was such a long time ago, ahaha.

I don't think this has anything to do with the protected fields stuff, maybe I just found a gap in the tests as the find CLP can never be `false`. Not sure. I should remove?

File: src/Controllers/DatabaseController.js (hunk 4 above)
Comment (awgeorge) [reply src/Controllers/DatabaseController.js]: That's correct, I was borrowing the naming convention from the previous function `addPointerPermissions`.

File: spec/UserPII.spec.js [hunk 9]
```diff
@@ -521,5 +521,699 @@ describe('Personally Identifiable Information', () => {
...
+
+      beforeEach(async done => {
+        const adminRole = await new Parse.Role(
+          'Administrator',
+          new Parse.ACL()
+        ).save(null, { useMasterKey: true });
+
+        const managementRole = new Parse.Role(
+          'managementOf_user' + user.id,
+          new Parse.ACL(user)
```
[the remark this answers is not in the document]
Comment (awgeorge) [reply spec/UserPII.spec.js]: From what I understand, passing a user here means that this management role can only be edited by the user in question. Does this default to the currently logged in user? Didn't find anything in the docs.

**Comment by awgeorge** [comment]
@acinader Thanks for reviewing, I should be able to sort those specs out, but I won't be able to get to it for a little while. Maybe if someone in the community who wanted this feature could lend a hand with the last tests that would be amazing. I'll pick it back up when my storm blows over if the work is still outstanding.

File: spec/UserPII.spec.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (acinader) [reply spec/UserPII.spec.js]: From the api docs: 'If no argument is given, the ACL has no permissions for anyone. If the argument is a Parse.User, the ACL will have read and write permission for only that user.'  

in any event, I was just asking....

**Comment by acinader** [comment]
@awgeorge there's a few failing tests for pii now.  hope it wasn't one of my suggestions :0.

we don't need to wait for the tests i'm asking for to merge this into master and I can try my hand at it but not until next week.

if someone can take a look at the failed test, otherwise I'll try to later.

File: src/Controllers/DatabaseController.js (hunk 3 above)
Comment (awgeorge) [reply src/Controllers/DatabaseController.js]: @acinader looks like this is the offending line. I'll reverse.

**Pull request merged**